### PR TITLE
Utc to local string converter

### PR DESCRIPTION
**Pull request by LeoJHarris**
### Description of Change ###

Adds a new converter for UTC (DateTime or DateTimeOffset) to local time string https://github.com/xamarin/XamarinCommunityToolkit/issues/1260. Optional DateTimeFormat can be provided otherwise fallback format available.

### PR Checklist ###

- [x] Has a linked Issue, and the Issue has been `approved`
- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [ ] Rebased on top of main at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Xamarin.CommunityToolkit.csproj [hunk 1]
```diff
@@ -66,6 +66,9 @@
     <None Include="..\..\..\assets\XamarinCommunityToolkit_128x128.png" PackagePath="icon.png" Pack="true" />
     <PackageReference Include="Xamarin.Forms" Version="5.0.0.2083" />
     <PackageReference Include="mdoc" Version="5.8.3" PrivateAssets="all" />
+    <Compile Include="Converters\UtcDateTimeToLocalStringConverter.shared.cs">
+      <SubType></SubType>
+    </Compile>
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Xamarin.CommunityToolkit.csproj:71]: Please remove this (all *.shared.cs files are added by rule on line 63)
```suggestion
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+		public string DateTimeFormat
+		{
+			get => (string)GetValue(DateTimeFormatProperty);
+			set => SetValue(DateTimeFormatProperty, value);
+		}
+
+		/// <summary>
+		/// Backing BindableProperty for the <see cref="DateTimeFormat"/> property.
+		/// </summary>
+		public static readonly BindableProperty DateTimeFormatProperty = BindableProperty.Create(nameof(DateTimeFormat), typeof(string), typeof(UtcDateTimeToLocalStringConverter), defaultValue: "g");
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs:27]: please move it at the top on class

Comment (LeoJHarris) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs:27]: @AndreiMisiukevich will do with the required changes shortly. I think it was the code maid extension that has reorganised the class.

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+			throw new ArgumentException("Value must be type DateTime or DateTimeOffset", nameof(value));
+		}
+
+		bool IsValidDateFormat(string dateFormat)
+		{
+			try
+			{
+				var s = DateTime.Now.ToString(dateFormat, CultureInfo.InvariantCulture);
+				DateTime.Parse(s, CultureInfo.InvariantCulture);
+				return true;
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs:57]: Can we check it anyhow else?
Raising exception is quite expensive operation

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+		/// </summary>
+		public static readonly BindableProperty DateTimeFormatProperty = BindableProperty.Create(nameof(DateTimeFormat), typeof(string), typeof(UtcDateTimeToLocalStringConverter), defaultValue: "g");
+
+		public object Convert(object? value, Type? targetType, object? parameter, CultureInfo? culture)
+		{
+			if (value is DateTime or DateTimeOffset)
+			{
+				if (!string.IsNullOrEmpty(DateTimeFormat))
+				{
+					if (DateTimeFormat is string dateTimeFormat && !IsValidDateFormat(DateTimeFormat))
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs:35]: DateTimeFormat is always a string, isn't it?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs (hunk 2 above)
Comment (AndreiMisiukevich) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs:27]: Btw, why not null by default?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs (hunk 4 above)
Comment (LeoJHarris) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs:35]: @AndreiMisiukevich Yes, proposing to remove `DateTimeFormat is string dateTimeFormat`? Yes we can remove this and keep the check against a valid DateTimeFormat... for example a date time format of "asd" would be invalid. Some unit tests check for this.

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+					return dateTime.ToLocalTime().ToString(DateTimeFormat);
+
+				if (value is DateTimeOffset dateTimeOffset)
+					return dateTimeOffset.UtcDateTime.ToLocalTime().ToString(DateTimeFormat);
+
+				return string.Empty;
+			}
+
+			throw new ArgumentException("Value must be type DateTime or DateTimeOffset", nameof(value));
+		}
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs:49]: This method can be rewritten as
```csharp
if (!IsValidDateFormat(DateTimeFormat))
	throw new ArgumentException("Value must be a valid date time format", nameof(DateTimeFormat));
	
if (value is DateTime dateTime)
	return dateTime.ToLocalTime().ToString(DateTimeFormat);

if (value is DateTimeOffset dateTimeOffset)
	return dateTimeOffset.UtcDateTime.ToLocalTime().ToString(DateTimeFormat);

throw new ArgumentException("Value must be type DateTime or DateTimeOffset", nameof(value));

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs (hunk 2 above)
Comment (LeoJHarris) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs:27]: @AndreiMisiukevich Null is fine, I personally like the default so as not to throw exception and provide a general date time format. But his can be changed to null and handle this appropriately.

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs (hunk 3 above)
Comment (LeoJHarris) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs:57]: @AndreiMisiukevich Good point `DateTime.Parse` will throw `FormatException`. We can use `DateTime.TryParse` which simply return true or false.

**Comment by LeoJHarris** [comment]
@AndreiMisiukevich Have updated with code changes as mentioned.

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	/// <summary>
+	/// Converts the incoming value to a <see cref="string"/> local time represenation in the
+	/// datetime format provided.
+	/// </summary>
+	public class UtcDateTimeToLocalStringConverter : ValueConverterExtension, IValueConverter
+	{
+		/// <summary>
+		/// Backing BindableProperty for the <see cref="DateTimeFormat"/> property.
+		/// </summary>
+		public static readonly BindableProperty DateTimeFormatProperty = BindableProperty.Create(nameof(DateTimeFormat), typeof(string), typeof(UtcDateTimeToLocalStringConverter), defaultValue: "g");
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs:17]:
```suggestion
		public static readonly BindableProperty DateTimeFormatProperty 
			= BindableProperty.Create(nameof(DateTimeFormat), typeof(string), typeof(UtcDateTimeToLocalStringConverter), defaultValue: "g");
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+		/// <param name="parameter">N/A</param>
+		/// <param name="culture">N/A</param>
+		/// <returns>N/A</returns>
+		public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) => throw new NotImplementedException();
+
+		private bool IsValidDateFormat(string dateFormat)
+		{
+			var s = DateTime.Now.ToString(dateFormat, CultureInfo.InvariantCulture);
+			return DateTime.TryParse(s, out _);
+		}
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs:57]:
```suggestion
			=> DateTime.TryParse(DateTime.UtcNow.ToString(dateFormat, CultureInfo.InvariantCulture), out _);
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+			if (!IsValidDateFormat(DateTimeFormat))
+				throw new ArgumentException("Value must be a valid date time format", nameof(DateTimeFormat));
+
+			if (value is DateTime dateTime)
+				return dateTime.ToLocalTime().ToString(DateTimeFormat);
+
+			if (value is DateTimeOffset dateTimeOffset)
+				return dateTimeOffset.UtcDateTime.ToLocalTime().ToString(DateTimeFormat);
+
+			throw new ArgumentException("Value must be type DateTime or DateTimeOffset", nameof(value));
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs:40]:
```suggestion
			throw new ArgumentException("Value must be of type DateTime or DateTimeOffset", nameof(value));
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+		/// <summary>
+		/// This method is not implemented and will throw a <see cref="NotImplementedException"/>.
+		/// </summary>
+		/// <param name="value">N/A</param>
+		/// <param name="targetType">N/A</param>
+		/// <param name="parameter">N/A</param>
+		/// <param name="culture">N/A</param>
+		/// <returns>N/A</returns>
+		public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) => throw new NotImplementedException();
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs:51]:
```suggestion
		public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) 
			=> throw new NotImplementedException();
```

**Comment by AndreiMisiukevich** [comment]
@jfversluis @brminnick is this required to post documentation for this converter?

**Review by AndreiMisiukevich — APPROVED** [review]

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+		/// </summary>
+		/// <param name="value">N/A</param>
+		/// <param name="targetType">N/A</param>
+		/// <param name="parameter">N/A</param>
+		/// <param name="culture">N/A</param>
+		/// <returns>N/A</returns>
+		public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) 
+			=> throw new NotImplementedException();
+
+		private bool IsValidDateFormat(string dateFormat)
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/UtcDateTimeToLocalStringConverter.shared.cs:55]:
```suggestion
		bool IsValidDateFormat(string dateFormat)
```

**Comment by jfversluis** [comment]
> @jfversluis @brminnick is this required to post documentation for this converter?

I see an issue has been created on the repo docs, ideally the documentation for it will be written by the author as well :)

**Review by jfversluis — APPROVED** [review]

**Comment by jfversluis** [comment]
Thanks for your contributions @LeoJHarris !

**Comment by LeoJHarris** [comment]
@jfversluis My Pleasure 👍

**Pull request merged**